### ISSUE TITLE
Android system back button or navigation back gesture always pops entire app

### ISSUE THREAD

**Issue by rydmike** [csells/go_router#57]
When popping routes that have been routed to as sub-routes or via push, it is **expected** that Android system back button or system navigation back swipe, would pop back to previous screen, until we are back on start screen, after which the app would be popped. However, in all **GoRouter** examples the Android back button or back navigation gesture, always directly pops the entire app.

This can be demonstrated e.g. with example `sub_routes.dart`:

<img src="https://user-images.githubusercontent.com/39990307/135772752-5400b783-d887-45d3-9f75-96c0c145a64d.gif" width="300">

or with example `push.dart`:

<img src="https://user-images.githubusercontent.com/39990307/135772755-23cbb3d2-dfb0-41ab-b9ba-ac1faad70bf2.gif" width="300">

---

I have not yet tried to figure out why this happens 😃

**Comment by csells** [comment]
yep. that's a bug alright. the Android Back button should work the same as the AppBar Back button.

**Comment by rydmike** [comment]
Actually Android back button and browser back do not and should not behave the same.

Web back is always chronological back, while back on app bar and and on Android system back button or back swipe gesture, is "always" up. But it is not that simple, to go down the rabbit hole, here is good read on the topic:

https://github.com/tomgilder/routemaster/issues/176#issuecomment-881859680

TLDR: It's a mess.

**Comment by masreplay** [comment]
> yep. that's a bug alright. the Android Back button should work the same as the AppBar Back button.

when we will get the fixes please

**Comment by esDotDev** [comment]
> Actually Android back button and browser back do not and should not behave the same.
> 
> Web back is always chronological back, while back on app bar and and on Android system back button or back swipe gesture, is "always" up. But it is not that simple, to go down the rabbit hole, here is good read on the topic:
> 
> [tomgilder/routemaster#176 (comment)](https://github.com/tomgilder/routemaster/issues/176#issuecomment-881859680)
> 
> TLDR: It's a mess.

Well back on the web is just a deeplink, inside a chronological array of deeplinks that the browser maintains, it doesn't really have any relation to System Back or a UI back button.

I think this is generally correct that Android Back should work the same as AppBar back, but we should also be able to intercept it, in case we have some fullscreen popup, or maybe a widget that has it's own internal "back" handler.
